tests/5.0/target/test_target_device.F90

```diff
@@ -0,0 +1,93 @@
+!===--- test_target_device.F90 -------------------------------===//
+!
+! OpenMP API Version 5.0 Nov 2018
+!
+! This test checks the target construct with device clause where device-
+! modifier is either ancestor or device_num. If no device_modifier is 
+! present, the behavior is the same as if device_num were present.
+! 
+!//===----------------------------------------------------------------------===//
+#include "ompvv.F90"
+
+#define N 1024
+
+! Required for device(ancestor: 1)
+!$omp requires reverse_offload
+
+PROGRAM test_target_device
+  USE iso_fortran_env
+  USE ompvv_lib
+  USE omp_lib
+  implicit none
+  OMPVV_TEST_OFFLOADING
+
+  OMPVV_TEST_VERBOSE(target_device_ancestor() .ne. 0)
+  OMPVV_TEST_VERBOSE(target_device_device_num() .ne. 0)
+
+  OMPVV_REPORT_AND_RETURN()
+
+CONTAINS
+  INTEGER FUNCTION target_device_ancestor()
+    INTEGER :: errors, i, which_device
+    INTEGER, DIMENSION(N) :: a
+
+    errors = 0
+    which_device = 0
+
+    DO i = 1, N
+       a(i) = i
+    END DO
+
+    OMPVV_TEST_AND_SET(errors, omp_get_num_devices() .le. 0) 
+    OMPVV_ERROR_IF(omp_get_num_devices() .le. 0, "Since no target devices were found, this test will be skipped.")
+
+    IF ( omp_get_num_devices() .gt. 0 ) THEN
+
+    !$omp target device(ancestor: 1) map(tofrom: a) map(to: which_device)
+    DO i = 1, N
+       a(i) = a(i) + 2
+    END DO
+    which_device = 75
```

Comment:
```suggestion
    which_device = omp_get_num_device()
```
Would it make sense to check that omp_get_device_num() returns the same number as omp_get_device_num() on the host?

Reply:
The problem with that is that compilers (or at least GCC) will reject this with:
`Error: OpenMP runtime API call ‘omp_get_device_num’ in a region with ‘device(ancestor)’ clause`
_(with ...num_device changes to ...device_num)_

That's the last item in the following quote from OpenMP 5.0, https://www.openmp.org/spec-html/5.0/openmpsu60.html#x86-2820002.12.5
>  * If a device clause in which the ancestor device-modifier appears is present on the construct, then the following restrictions apply:
>    * A requires directive with the reverse_offload clause must be specified;
>    * The device clause expression must evaluate to 1;
>    * Only the device, firstprivate, private, defaultmap, and map clauses may appear on the construct;
>    * No OpenMP constructs or calls to OpenMP API runtime routines are allowed inside the corresponding target region.
 * * *
However, I also have an issue with "= 75" – with unified shared memory, this likely also gets assigned to 75 even if the reverse lookup works. Thus, this will give spurious FAILS.

I think it makes sense to call some procedure from that target region – and use that as better check:
* Not fully legal, but you could smuggle in an `omp_is_initial_device()` or `omp_get_device_num()` into that procedure as a compiler usually won't diagnose this and those API routines are likely to work correctly. — Downside: Not really legal, even if it likely works.
* The call to some procedure also is a check by itself – as this will give a link-time error if it is called from the device but only has a host version. There are two ways:
   * Write a procedure yourself. — Downside: If the procedure is in the same file, compilers are supposed to auto-tag them as 'declare target' for non-reverse offload. But if the compiler has a bug, it might still do it also for 'ancestor:1'.
   * Call some library function which is known to be available on the host (POSIX function) but unlikely to exist on the device. An example could be `v = getenv ("PATH")` – and check the env var exists (`v != NULL`) and has non-zero length (`v[0] != '\0'`); in Fortran, that would be: `call get_environment_variable ("PATH", length=len)` with `len > 0` check. — Could also be another procedure, which is likely supported on the host but not on the device.

Reply:
@tob2 As you said, there will be an issue `which_device = 75` if the system supports unified shared memory, and thank you for suggesting alternative solutions, but none of which seems to be guaranteed to work.